src/hooks/useActiveHash.js

```diff
@@ -0,0 +1,40 @@
+import { useState, useEffect } from "react"
+
+/**
+ * A hook to determine which section of the page is currently in the viewport.
+ * @param {*} itemIds Array of document ids to observe
+ * @param {*} rootMargin
+ * @returns id of the element currently in viewport
+ */
+export const useActiveHash = (itemIds, rootMargin = undefined) => {
+  const [activeHash, setActiveHash] = useState(``)
+
+  useEffect(() => {
+    const observer = new IntersectionObserver(
+      (entries) => {
+        entries.forEach((entry) => {
+          if (entry.isIntersecting) {
+            setActiveHash(`#${entry.target.id}`)
+          }
+        })
+      },
+      { rootMargin: rootMargin || `0% 0% -80% 0%` }
```

Comment:
Why the `-80%` is needed?

Reply:
Lets set the `0% 0% -80% 0%` value as the default arg value.

```js
export const useActiveHash = (itemIds, rootMargin = `0% 0% -80% 0%`) => {
  ...
}
```

Reply:
Using rootMargin we can grow or shrink the root element. The callback is fired when the observing element interacts with the root element (browser viewport in this case). 

By setting the bottom margin to -80% we ensure that the callback is fired when the heading is in the top 20% of the browser viewport. You can observe that the highlight in table of content doesn't change when the heading is visible. It changes when the heading is scrolled to the top of the viewport.

Reply:
> Lets set the 0% 0% -80% 0% value as the default arg value.

Done 👍🏽

Reply:
Oh I see, thanks for the explanation!